### PR TITLE
Create basic ViT model

### PR DESCRIPTION
**Pull request by davidmascharka**
Add a basic vision transformer (ViT) model for classification. Sample definition for training on MNIST:
```python
>>> model = ViT(
...     input_shape=(28, 28),
...     input_dim=1,
...     patch_shape=(4, 4),
...     feature_dim=128,
...     num_classes=10,
...     num_transformer_blocks=6,
... )

>>> print(f"{sum(p.size for p in model.parameters):,} parameters")
403,072 parameters
```

I trained for 3 epochs with a batch size of 32.

After training for ~15 minutes per epoch for 2-3 epochs and testing our your CPU coolant :skull_and_crossbones: you'll have a model that gets ~95-97% accuracy. Sample loss plot:
image

### CODE REVIEW

File: examples/vision_transformer.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
+import mygrad as mg
+from mynn.initializers import glorot_normal
```
Comment (davidmascharka) [inline examples/vision_transformer.py:2]: initialization matters quite a lot. you can either do this, or use
```python
mg.random.randn(...) * 0.02  # or some other small value
```

File: examples/vision_transformer.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+    Returns
+        mg.Tensor, shape=(B, num_tokens, patch_width * patch_height * C), the patchified input.
+    """
+    b, c, h, w = images.shape
+    h_prime = h // patch_height
+    w_prime = w // patch_width
+    images = images.reshape(b, c, h_prime, patch_height, w_prime, patch_width)
+    images = images.transpose(0, 2, 4, 3, 5, 1)
+    return images.reshape(b, h_prime * w_prime, patch_height * patch_width * c)
```
Comment (davidmascharka) [inline examples/vision_transformer.py:36]: if you don't mind an extra dependency, [einops](https://einops.rocks/) makes this look much cleaner:
```python
return einops.rearrange(
    images, "b c (h p1) (w p2) -> b (h w) (p1 p2 c)", p1=patch_height, p2=patch_width
)
```

File: examples/vision_transformer.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+
+class MultiHeadAttention:
+    """
+    Basic Multi-head attention mechanism as described in https://arxiv.org/abs/1706.03762v7.
+
+    Notes:
+        - Uses pre-norm on the input.
+        - Does not apply dropout to the attention mask.
+        - Applies no masking to the attention mask (not causal attention/we can attend to future tokens)
```
Comment (davidmascharka) [inline examples/vision_transformer.py:60]: this is all you need if you want to train a sequence-to-sequence model

File: examples/vision_transformer.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        q = q.reshape(x.shape[0], x.shape[1], self._num_heads, -1).transpose(0, 2, 1, 3)
+
+        k = qkv[..., x.shape[-1]:-x.shape[-1]]
+        k = k.reshape(x.shape[0], x.shape[1], self._num_heads, -1).transpose(0, 2, 1, 3)
+
+        v = qkv[..., -x.shape[-1]:]
+        v = v.reshape(x.shape[0], x.shape[1], self._num_heads, -1).transpose(0, 2, 1, 3)
+
+        # (b, h, t, h_c) @ (b, h, h_c, t) -> (b, h, t, t) gives us the inter-token attention matrix
+        attn = mg.nnet.softmax((q @ k.transpose(0, 1, 3, 2)) * self._scale, axis=-1)
```
Comment (davidmascharka) [inline examples/vision_transformer.py:85]: just apply the mask before the softmax here:
```python
attn = q @ k.transpose(0, 1, 3, 2)
mask_indices = np.triu_indices(attn.shape[-1], k=1)
attn[..., *mask_indices] = float("-inf")
attn = mg.nnet.softmax(attn)
```
to get the causal version

**Comment by davidmascharka** [comment]
A bit curious to implement https://arxiv.org/abs/2404.02258 and see what effect it has on training time. Would be nice to not destroy my CPU if I can avoid it

**Comment by davidmascharka** [comment]
> A bit curious to implement https://arxiv.org/abs/2404.02258 and see what effect it has on training time. Would be nice to not destroy my CPU if I can avoid it

Hey that really works. My epoch time dropped from ~15 minutes down to ~7-8 minutes using the 12.5% value they cite as optimal. That only gives us 6x 4x4 patches of MNIST, which is actually pretty surprising that it's enough context to do well on MNIST. Getting 93.4% accuracy after the first epoch.

Messy version of what you'd need to change for MoD transformer here:
```python
class TransformerBlock:
    def __init__(self, feature_dim: int, enable_routing: bool = False, routing_pct: float = 0.125):
        self._attn = MultiHeadAttention(feature_dim)
        self._mlp = MLP(feature_dim)
        if enable_routing:
            self._router = glorot_normal(feature_dim, 1)
            self._routing_pct = routing_pct
        self._routing = enable_routing

    def __call__(self, x: mg.Tensor) -> mg.Tensor:
        if self._routing:
            router_weights = (x @ self._router).squeeze()
            k = int(self._routing_pct * x.shape[1])
            top_k_indices = np.argpartition(router_weights.data, kth=-k)[:, -k:]
            router_index = (np.arange(len(top_k_indices))[:, np.newaxis], top_k_indices)
            selected_tokens = x[router_index]
        else:
            selected_tokens = x

        attn_out = self._attn(selected_tokens) + selected_tokens
        mlp_out = self._mlp(attn_out) + attn_out
        if self._routing:
            mlp_out = mlp_out * router_weights[router_index]
            x[router_index] = mlp_out
        else:
            x = mlp_out
        return x


class ViT:
    def __init__(
        self,
        input_shape: tuple[int, int],
        input_dim: int, 
        patch_shape: tuple[int, int],
        feature_dim: int,
        num_classes: int,
        num_transformer_blocks: int = 6,
        enable_routing: bool = False,
    ):
        ...
        self._transformer_blocks = tuple(
            TransformerBlock(feature_dim, enable_routing=enable_routing and i%2 == 1)
            for i in range(num_transformer_blocks)
        )
```

Note that for the causal version you'd need to add the auxiliary classifier and loss to predict which tokens will be top-k.
